srml/contracts/src/wasm/runtime.rs

```diff
@@ -586,6 +586,86 @@ define_env!(Env, <E: Ext>,
 		Ok(())
 	},
 
+	// Attempt to restore the specified contract.
+	//
+	// This will compute a tombstone hash from the caller's storage and the given code hash and
+	// if the hash matches the hash found in the tombstone at the specified address - kill the
+	// caller contract and restore the destination contract and set the specified `rent_allowance`.
+	// All caller's funds are transfered to the destination.
+	//
+	// This function doesn't perform restoration right away but defers it to the end of the
+	// transaction. If there is no tombstone in the destination address or if the hashes don't match
+	// then restoration is cancelled and no changes are made.
+	//
+	// `dest_ptr`, `dest_len` - the pointer and the length of a buffer that encodes `T::AccountId`
+	// with the address of the to be restored contract.
+	// `code_hash_ptr`, `code_hash_len` - the pointer and the length of a buffer that encodes
+	// a code hash of the to be restored contract.
+	// `rent_allowance_ptr`, `rent_allowance_len` - the pointer and the length of a buffer that
+	// encodes the rent allowance that must be set in the case of successful restoration.
+	// `delta_ptr` is the pointer to the start of a buffer that has `delta_count` storage keys
+	// laid out sequentially.
+	ext_restore_to(
+		ctx,
+		dest_ptr: u32,
+		dest_len: u32,
+		code_hash_ptr: u32,
+		code_hash_len: u32,
+		rent_allowance_ptr: u32,
+		rent_allowance_len: u32,
+		delta_ptr: u32,
+		delta_count: u32
+	) => {
+		let dest = {
+			let dest_buf = read_sandbox_memory(ctx, dest_ptr, dest_len)?;
+			<<E as Ext>::T as system::Trait>::AccountId::decode(&mut &dest_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let code_hash = {
+			let code_hash_buf = read_sandbox_memory(ctx, code_hash_ptr, code_hash_len)?;
+			<CodeHash<<E as Ext>::T>>::decode(&mut &code_hash_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let rent_allowance = {
+			let rent_allowance_buf = read_sandbox_memory(
+				ctx,
+				rent_allowance_ptr,
+				rent_allowance_len
+			)?;
+			BalanceOf::<<E as Ext>::T>::decode(&mut &rent_allowance_buf[..])
+				.ok_or_else(|| sandbox::HostError)?
+		};
+		let delta = {
+			// We don't use `with_capacity` here to not eagerly allocate the user specified amount
```

Comment:
Hmm, why not? Couldn't you just do an up-front bounds check on `delta_ptr` and `delta_ptr + KEY_SIZE * delta_count` then reserve the correct amount of memory?

Reply:
I was more thinking about if the user specifies the count `(usize)-1` the runtime will happily try to allocate this amount of memory and then abort due to OOM. Here we do this incrementally with paying as we go.

Changing this would require an explicit gas charge and and corresponding alterations. The optimization makes sense to me, but this is just something that I don't want to do just yet.